container/src/main/java/com/flipkart/poseidon/async/PoseidonConsumer.java

```diff
@@ -43,17 +59,94 @@ public PoseidonConsumer(Application application, Configuration configuration) {
+        }
+
+        HystrixRequestContext hystrixRequestContext = HystrixRequestContext.initializeContext();
+        initAllContext(request);
+
+        PoseidonResponse response = null;
         try {
-            PoseidonResponse response = new PoseidonResponse();
+            response = new PoseidonResponse();
             this.application.handleRequest(request, response);
             return new AsyncConsumerResult(AsyncResultState.SUCCESS);
         } catch (Throwable throwable) {
             logger.error("Unexpected exception while consuming async event", throwable);
             return new AsyncConsumerResult(AsyncResultState.FAILURE);
+        } finally {
+            ingestResponseBasedMetrics(response);
+            logFailedHystrixCommands(consumerRequest);
+            shutdownAllContext(hystrixRequestContext);
         }
     }
+
+    private void initAllContext(PoseidonRequest request) {
+        RequestContext.initialize();
+        ServiceContext.initialize(configuration.getResponseHeadersToCollect());
+        setContext(request);
+    }
+
+    private void setContext(PoseidonRequest request) {
+        RequestContext.set(METHOD, request.getAttribute(METHOD).toString());
+        RequestContext.set(SOURCE_ADDRESS, request.getUrl());
+
+        if (configuration.getHeadersConfiguration() != null && configuration.getHeadersConfiguration().getGlobalHeaders() != null) {
+            Map<String, String> headers = new HashMap<>();
+            for (HeaderConfiguration headerConfiguration : configuration.getHeadersConfiguration().getGlobalHeaders()) {
+                String value = request.getHeader(headerConfiguration.getName());
+                if (value == null) {
+                    value = headerConfiguration.getDefaultValue();
+                }
+                if (value != null) {
+                    headers.put(headerConfiguration.getName(), value);
+                }
+            }
+
+            ImmutableMap<String, String> immutableHeaders = ImmutableMap.copyOf(headers);
+            ServiceContext.set(ServiceClientConstants.HEADERS, immutableHeaders);
+            ServiceContext.set(ServiceClientConstants.COMMANDS, new ConcurrentLinkedQueue<String>());
+            ServiceContext.set(ServiceClientConstants.COLLECT_COMMANDS, configuration.collectServiceClientCommandNames());
+            ServiceContext.set(ServiceClientConstants.THROW_ORIGINAL, configuration.throwOriginalExceptionsForNonUpstreamFailures());
+            RequestContext.set(HEADERS, immutableHeaders);
+            MDC.setContextMap(immutableHeaders);
+        }
+    }
+
+    private void shutdownAllContext(HystrixRequestContext hystrixRequestContext) {
+        RequestContext.shutDown();
+        ServiceContext.shutDown();
+        hystrixRequestContext.shutdown();
+        MDC.clear();
+    }
+
+    private void ingestResponseBasedMetrics(PoseidonResponse response) {
+        if (!StringUtils.isNullOrEmpty(RequestContext.get(ENDPOINT_NAME))) {
```

Comment:
Is ENDPOINT_NAME set in request context in async flow?

Reply:
yes, it should be. its set when we look for the buildable in APILegoset